Lib/test/test_capi/test_codecs.py

```diff
@@ -516,5 +521,230 @@ def test_asrawunicodeescapestring(self):
+        self.assertRaises(LookupError, self.codecs.lookup, self.encoding_name)
+
+    def test_codec_register(self):
+        search_function, encoding = self.search_function, self.encoding_name
+        self.assertIsNone(_testcapi.codec_register(search_function))
+        self.assertIs(self.codecs.lookup(encoding), search_function(encoding))
+        self.assertEqual(self.codecs.encode('123', encoding=encoding), '321')
+
+    def test_codec_unregister(self):
+        search_function, encoding = self.search_function, self.encoding_name
+        self.assertRaises(LookupError, self.codecs.lookup, encoding)
+        self.codecs.register(search_function)
+        self.assertIsNone(_testcapi.codec_unregister(search_function))
+        self.assertRaises(LookupError, self.codecs.lookup, encoding)
+
+    def test_codec_known_encoding(self):
+        self.assertRaises(LookupError, self.codecs.lookup, 'unknown-codec')
+        self.assertFalse(_testcapi.codec_known_encoding('unknown-codec'))
+        self.assertFalse(_testcapi.codec_known_encoding('unknown_codec'))
+        self.assertFalse(_testcapi.codec_known_encoding('UNKNOWN-codec'))
+
+        encoding_name = self.encoding_name
+        self.assertRaises(LookupError, self.codecs.lookup, encoding_name)
+        self.codecs.register(self.search_function)
+
+        for name in [
+            encoding_name,
+            encoding_name.upper(),
+            encoding_name.replace('_', '-'),
+        ]:
+            with self.subTest(name):
+                self.assertTrue(_testcapi.codec_known_encoding(name))
+
+    def test_codec_encode(self):
+        encode = _testcapi.codec_encode
+        self.assertEqual(encode('a', 'utf-8', NULL), b'a')
+        self.assertEqual(encode('a', 'utf-8', 'strict'), b'a')
+        self.assertEqual(encode('é', 'ascii', 'ignore'), b'')
+        # todo: add more cases
+        self.assertRaises(TypeError, encode, NULL, 'ascii', 'strict')
+        # CRASHES encode('a', NULL, 'strict')
+
+    def test_codec_decode(self):
+        decode = _testcapi.codec_decode
+
+        b = b'a\xc2\xa1\xe4\xbd\xa0\xf0\x9f\x98\x80'
+        s = 'a\xa1\u4f60\U0001f600'
+
+        self.assertEqual(decode(b, 'utf-8', 'strict'), s)
+        self.assertEqual(decode(b, 'utf-8', NULL), s)
+        self.assertEqual(decode(b, 'latin1', 'strict'), b.decode('latin1'))
+        self.assertRaises(UnicodeDecodeError, decode, b, 'ascii', 'strict')
+        self.assertRaises(UnicodeDecodeError, decode, b, 'ascii', NULL)
+        self.assertEqual(decode(b, 'ascii', 'replace'), 'a' + '\ufffd'*9)
+        # todo: add more cases
+
+        # _codecs.decode only reports unknown errors policy when they are
+        # used (it has a fast path for empty bytes); this is different from
+        # PyUnicode_Decode which checks that both the encoding and the errors
+        # policy are recognized.
+        self.assertEqual(decode(b'', 'utf-8', 'unknown-errors-policy'), '')
```

Comment:
The error handler should be checked if Python is built in debug mode, or if -X dev is used, no? I don't recall when it's checked.

Reply:
Ah I forgot that comment. I'll address it in a moment.

Reply:
Ok, my comment was a bit misleading:

- codecs.decode() will not check the error policy if it's not needed
- bytes.decode() checks the error policy *before* even trying to actually decode the content.

I added this test to reflect this behaviour but I said that there was a fast path for empty bytes, which I think it is not the case. So there is nothing to do with the debug mode I think (but perhaps I'm missing something?)